### PR TITLE
Add sql text counts stats to `vtcombo`,`vtgate`+`vttablet`

### PR DESCRIPTION
**Pull request by timvaillancourt**
## Description

This PR adds stats for the size of query text processed by `vtgate` and `vttablet`

This is useful to understand large queries that potentially cause performance, grpc message size, bandwidth problems

## Related Issue(s)

Resolves https://github.com/vitessio/vitess/issues/15929

## Checklist

-   [x] "Backport to:" labels have been added if this change should be back-ported to release branches
-   [x] If this change is to be back-ported to previous releases, a justification is included in the PR description
-   [x] Tests were added or are not required
-   [x] Did the new or modified tests pass consistently locally and on CI?
-   [ ] Documentation was added or is not required

### CODE REVIEW

**Comment by timvaillancourt** [comment]
Opened this one with a lot of failing CI to see if others have ideas how to debug

The failures are confusing and feel unrelated to this change 🤔

**Comment by harshit-gangal** [comment]
> Opened this one with a lot of failing CI to see if others have ideas how to debug
> 
> The failures are confusing and feel unrelated to this change 🤔

There failures are related to the changes made. I have pushed a commit that fixes some of those. You should run the test locally and fix the remaining if any.

**Comment by timvaillancourt** [comment]
> > Opened this one with a lot of failing CI to see if others have ideas how to debug
> > The failures are confusing and feel unrelated to this change 🤔
> 
> There failures are related to the changes made. I have pushed a commit that fixes some of those. 

@harshit-gangal indeed it is, thanks! 🤦 🙇 

> You should run the test locally and fix the remaining if any.

Good point, I've made a bad habit of relying on GitHub Actions CI because many `go test` CIs fail on my MacBook when they rely on external things (perhaps not this test). I'll give running CI locally another shot 👀 

Somehow this was not obvious from the CI outputs I looked at, including [this `Unit test` run]( https://github.com/vitessio/vitess/actions/runs/9064779344/job/24903791921) that ends in just:
```bash
+ go-junit-report -set-exit-code
make: *** [Makefile:209: unit_test] Error 1
Error: Process completed with exit code 1.
```

Perhaps it's just the v15 fork/release I'm working on, but I'm used to the `Unit test` CI including the output of `go test -v` to help explain the failure. I'll see if something about that changed and if there is an alternative

**Comment by harshit-gangal** [comment]
> Perhaps it's just the v15 fork/release I'm working on, but I'm used to the Unit test CI including the output of go test -v to help explain the failure. I'll see if something about that changed and if there is an alternative

The first step records the output, if you click on the step below in the CI you will see the output.

**Comment by harshit-gangal** [comment]
Screenshot 2024-05-17 at 8 54 54 PM

**Comment by timvaillancourt** [comment]
> > Perhaps it's just the v15 fork/release I'm working on, but I'm used to the Unit test CI including the output of go test -v to help explain the failure. I'll see if something about that changed and if there is an alternative
> 
> The first step records the output, if you click on the step below in the CI you will see the output.

@harshit-gangal ahh, that's where it goes! Thanks a lot, more on this one soon 👀

File: go/vt/vtgate/vtgate.go
Comment (deepthi) [inline go/vt/vtgate/vtgate.go:214]: How does `/metrics` render this one? If it turns out to be vtgate_s_q_l_text_counts, we'll want to rename it :)
In general, for any metrics changes, it is good to post screenshots or embedded text from both `/debug/vars` and `/metrics` into the PR description. That serves as a visual check for what we are emitting.

**Review by deepthi — APPROVED** [review]
Looks like a good change. Approving with the caveat that the question about the metric name should be addressed. We'll need one more review from @vitessio/query-serving

**Comment by deepthi** [comment]
I removed the website docs label. We don't actually have a page that lists all available metrics, so there's no good place to add docs for new metrics.

**Review by frouioui — APPROVED** [review]

**Comment by frouioui** [comment]
Since the PR has two approvals but an outstanding question I am adding the `Do Not Merge` label so we don't merge this by accident.

**Comment by timvaillancourt** [comment]
@deepthi / @frouioui, here are the metric outputs:
```bash
tvaillancourt@REDACTED:~$ curl -s localhost:15000/debug/vars | jq .QuerySQLTextCounts
{
  "Join.Select.": 3234,
  "dual.Select.": 3615
}
```
```bash
tvaillancourt@REDACTED:~$ curl -s localhost:15000/metrics 2>&1 | grep vttablet_query_sql_text_counts
# HELP vttablet_query_sql_text_counts query sql text counts
# TYPE vttablet_query_sql_text_counts counter
vttablet_query_sql_text_counts{plan="Select",table="Join",workload=""} 3234
vttablet_query_sql_text_counts{plan="Select",table="dual",workload=""} 3650
```

So looks good 👍. While doing that I thought of two potential alternatives in case this didn't work: `QueryTextProcessed` or `QueryTextRead`, any thoughts?

**Review by shlomi-noach — APPROVED** [review]
Somewhat repeating @frouioui , will merge once @deepthi 's concern is addressed (likely you need o change `...SQL...` to `...Sql...` and will then merge.

**Comment by timvaillancourt** [comment]
> Somewhat repeating @frouioui , will merge once @deepthi 's concern is addressed (likely you need o change `...SQL...` to `...Sql...` and will then merge.

@shlomi-noach the potential issue of `s_q_l` that Deepthi pointed out didn't seem to be a problem in practice, here is the metric output:
```bash
tvaillancourt@REDACTED:~$ curl -s localhost:15000/metrics 2>&1 | grep vttablet_query_sql_text_counts
# HELP vttablet_query_sql_text_counts query sql text counts
# TYPE vttablet_query_sql_text_counts counter
vttablet_query_sql_text_counts{plan="Select",table="Join",workload=""} 3234
vttablet_query_sql_text_counts{plan="Select",table="dual",workload=""} 3650
```

Let me know if any change is needed. I'll add a `changelog/` note once we agree on the metric name and I think we're done! 🙇

File: go/vt/vtgate/vtgate.go
Comment (deepthi) [reply go/vt/vtgate/vtgate.go:214]: OK, now I understand the intent of this metric, it won't work as implemented. Counters should only be used for monotonically increasing numbers, prometheus will not let you reduce the value of a "counter".
The right type of metric to use here is a Gauge, which can take any value, and change up or down. If I had read the linked issue properly, I would have realized this during the original review.
The name of the metric is also misleading, because it is not a "count" of some specific kind of sql text, it is instead the length in text of the query that is being processed, so the name needs to reflect that.

Comment (deepthi) [reply go/vt/vtgate/vtgate.go:214]: `QueryTextProcessed` will in fact be a better name for the metric.

**Review by deepthi — CHANGES_REQUESTED** [review]

**Comment by deepthi** [comment]
@shlomi-noach please feel free to dismiss my review if @timvaillancourt is able complete the requested changes before code freeze.

**Comment by timvaillancourt** [comment]
> OK, now I understand the intent of this metric, it won't work as implemented. Counters should only be used for monotonically increasing numbers, prometheus will not let you reduce the value of a "counter".

@deepthi the intent of this metric is a counter that increments forever so we can infer the rate of query text processed at any given time. I don't believe there is a need to ever reduce this metric, it only increases from the time the daemon starts like other counters

File: go/vt/vtgate/vtgate.go
Comment (shlomi-noach) [reply go/vt/vtgate/vtgate.go:214]: The way I understand it, the metric would be more appropriately named `QueryProcessedTextLengths`

**Comment by shlomi-noach** [comment]
@timvaillancourt to summarize the current state: the changes make sense, and the metric looks correct. But if you could please rename it to something such as `QueryProcessedTextLengths`? And with that we can merge the PR.

**Comment by timvaillancourt** [comment]
> But if you could please rename it to something such as QueryProcessedTextLengths

@shlomi-noach / @deepthi I felt `Lengths` so I've gone with `QueryTextCharactersProcessed` which I think is long but very clear. I've also added this to the 20.0.0 changelog

Rerequested your review 🙇

File: changelog/20.0/20.0.0/summary.md
Comment (deepthi) [inline changelog/20.0/20.0.0/summary.md:331]:
```suggestion
VTGate and VTTablet expose a new counter stat `QueryTextCharactersProcessed` to reflect the number of query text characters processed.
```

File: changelog/20.0/20.0.0/summary.md
Comment (deepthi) [inline changelog/20.0/20.0.0/summary.md:333]: What is DbType? Looking at how `statsKey` is constructed, this seems to be TabletType. We use TabletType everywhere in the docs, so it would be confusing to add a new term.

Comment (deepthi) [inline changelog/20.0/20.0.0/summary.md:333]:
```suggestion
VTGate groups this metric by Operation, Keyspace and TabletType. On VTTablet it is grouped by Table, Plan and optionally Workload.
```

**Comment by deepthi** [comment]
Some minor wording changes to release notes are required to avoid user confusion. Rest LGTM. @shlomi-noach you may dismiss my review once these changes are done.

**Pull request merged**
